### PR TITLE
WIP: IaC inventory (for GitHub)

### PR DESCRIPTION
**Pull request by Eriner**
Do not merge.

### CODE REVIEW

File: cmd/iacinventorycmd/iacinventorycmd.go [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+			opts.GetPrinter().PrintResult(&buf, result)
+			if !submit {
+				// TODO: also early exit if auth is not configured
+				return nil
+			}
+			values := make(map[string]string) // TODO: add debugging values?
+			c := opts.GetAPIClient().GetClient()
+			req := c.R()
+			req.SetFileReader("file_iac_inventory", "iac_inventory.json", bytes.NewReader(buf.Bytes()))
+			req.SetHeader("X-SOLUBLE-ORG-ID", opts.GetAPIClientConfig().Organization)
```
Comment (slshen) [inline cmd/iacinventorycmd/iacinventorycmd.go:41]: Use XCPWithFileReader after I merge #80

File: cmd/iacinventorycmd/iacinventorycmd.go [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+				User:       username,
+				OauthToken: oauthToken,
+			})
+			result, err := scanner.Run()
+			if err != nil {
+				return err
+			}
+			opts.PrintResult(result)
+			var buf bytes.Buffer
+			opts.OutputFormat = "json" // is there a cleaner way to do this?
```
Comment (slshen) [inline cmd/iacinventorycmd/iacinventorycmd.go:31]: Yeah, just json.Marshal().  opts.PrintResult is for printing results out to the user.

File: pkg/iacinventory/ci.go [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+	CIGithub    CI = "github"
+	CIGitlab    CI = "gitlab"
+	CICircle    CI = "circle"
+	CIDrone     CI = "drone"
+	CIJenkins   CI = "jenkins"
+	CIAzure     CI = "azure"
+)
+
+// Test a given CI system against a file for a match, implements filepath.WalkFunc
+func WalkCI(_ string, info os.FileInfo, err error) (CI, error) {
```
Comment (slshen) [inline pkg/iacinventory/ci.go:21]: Don't export this

File: pkg/iacinventory/iacinventory.go [hunk 4]
```diff
@@ -0,0 +1,12 @@
+package iacinventory
+
+import "github.com/soluble-ai/go-jnode"
+
+type IacInventorier interface {
```
Comment (slshen) [inline pkg/iacinventory/iacinventory.go:5]: If it's GithubScanner is a IacInventorier, then it should be IacInventoryScanner I suppose.

File: pkg/iacinventory/repo.go [hunk 5]
```diff
@@ -0,0 +1,9 @@
+package iacinventory
+
+// Repo is returned by the inventory to be sent to the API as JSON array.
+type Repo struct {
+	Name               string   `json:"name"`
+	CIs                []string `json:"cis"`
```
Comment (slshen) [inline pkg/iacinventory/repo.go:6]: cis isn't a great name.  CISystems ?

File: pkg/iacinventory/iacinventory.go [hunk 6]
```diff
@@ -0,0 +1,12 @@
...
+	Run() (*jnode.Node, error)
+	Stop() error
```
Comment (slshen) [inline pkg/iacinventory/iacinventory.go:7]: What's Stop() for?  If nothing, then remove it

File: pkg/iacinventory/iacinventory.go [hunk 7]
```diff
@@ -0,0 +1,12 @@
+package iacinventory
+
+import "github.com/soluble-ai/go-jnode"
+
+type IacInventorier interface {
+	Run() (*jnode.Node, error)
```
Comment (slshen) [inline pkg/iacinventory/iacinventory.go:6]: Better to return a ([]Repo, error) and leave marshalling to the caller

File: pkg/iacinventory/github_repo.go [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+	}
+	defer resp.Body.Close()
+
+	// we stuff the extracted tarball into a temporary directory
+	dir, err := ioutil.TempDir("", "soluble-iac-scan")
+	if err != nil {
+		return fmt.Errorf("error creating temporary directory: %w", err)
+	}
+	g.dir = dir
+	gunzip, err := gzip.NewReader(resp.Body)
```
Comment (slshen) [inline pkg/iacinventory/github_repo.go:123]: Let's move this gunzip/untar code to a new package and include a test case.

We would add unzip code there as well.

File: pkg/iacinventory/github_repo.go [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+				}
+				return fmt.Errorf("error creating directory for tarball extraction: %w", err)
+			}
+		case tar.TypeReg:
+			// cut out the outer directory from the filename to ensure we add the files to g.dir (and not g.dir/some-random-subdir)
+			f, err := os.Create(outfile)
+			if err != nil {
+				return fmt.Errorf("error creating file during tarball extraction: %w", err)
+			}
+			c, err := io.CopyN(f, t, 100<<(10*2)) // 100MB max size
```
Comment (slshen) [inline pkg/iacinventory/github_repo.go:162]: This seems, big?  Surely we'd not be interested in files larger than even 1MB?

If refactoring this as a function this value should be a parameter or option to that function.

File: pkg/iacinventory/github_repo.go [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+					f.Close()
+					return fmt.Errorf("error writing file during tarball extraction: %w", err)
+				}
+			}
+			if c == 100<<(10*2) {
+				log.Infof("repository %q was larger than 100MB - skipping", g.FullName)
+				f.Close()
+				continue
+			}
+			f.Close()
```
Comment (slshen) [inline pkg/iacinventory/github_repo.go:174]: defer f.Close() after os.Create(), and wrap in func() { ... }() is better

File: pkg/iacinventory/github_repo.go [hunk 11]
```diff
@@ -0,0 +1,184 @@
...
+				return fmt.Errorf("error creating file during tarball extraction: %w", err)
+			}
+			c, err := io.CopyN(f, t, 100<<(10*2)) // 100MB max size
+			if err != nil {
+				if !errors.Is(err, io.EOF) {
+					f.Close()
+					return fmt.Errorf("error writing file during tarball extraction: %w", err)
+				}
+			}
+			if c == 100<<(10*2) {
```
Comment (slshen) [inline pkg/iacinventory/github_repo.go:169]: This would just detect a single file was at least 100MB.

File: pkg/iacinventory/github.go [hunk 12]
```diff
@@ -0,0 +1,241 @@
...
+	user, err := user.Current()
+	if err != nil {
+		retErr = fmt.Errorf("unable to get OS user: %w", err)
+		return
+	}
+
+	// Github credentials can exist in one of two files: ~/.config/gh/hosts.yml,
+	// or ~/.config/gh/config.yml. Creds _should_ only exist in the latter, but
+	// on some systems it appears in the former for whatever reason.
+	ghConfigFile = filepath.Join(user.HomeDir, ".config/gh/hosts.yml")
```
Comment (slshen) [inline pkg/iacinventory/github.go:178]: I think https://github.com/mitchellh/go-homedir is preferred, though to be honest I'm not completely sure.

File: pkg/iacinventory/github.go [hunk 13]
```diff
@@ -0,0 +1,241 @@
...
+		return
+	}
+	if err := yaml.Unmarshal(yamlf, &ghConfig); err != nil {
+		retErr = fmt.Errorf("unable to parse gh config: %w", err)
+		return
+	}
+
+	// Depending on the "type" of gh configuration (see file selector above),
+	// the yaml will be structured differently
+	if conf := ghConfig["github.com"]; conf != nil {
```
Comment (slshen) [inline pkg/iacinventory/github.go:208]: You might prefer using jnode for this: n := jnode.FromMap(ghConfig)

Then n.Path("github.com").Path("user").GetString() etc

File: pkg/iacinventory/github_repo.go [hunk 14]
```diff
@@ -0,0 +1,184 @@
...
+	"io/ioutil"
+	"net/http"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/soluble-ai/soluble-cli/pkg/log"
+)
+
+type GithubRepo struct {
```
Comment (slshen) [inline pkg/iacinventory/github_repo.go:21]: This seems to duplicate the Repo type?

File: pkg/iacinventory/github_repo.go [hunk 15]
```diff
@@ -0,0 +1,184 @@
...
+		scanner := bufio.NewScanner(f)
+		for scanner.Scan() {
+			if bytes.Contains(scanner.Bytes(), []byte("AWSTemplateFormatVersion")) {
+				if g.CloudformationDirs == nil {
+					g.CloudformationDirs = make(map[string]bool)
+				}
+				g.CloudformationDirs[strings.TrimPrefix(filepath.Dir(filepath.Clean(path)), g.dir+"/")] = true
+			}
+		}
+		f.Close()
```
Comment (slshen) [inline pkg/iacinventory/github_repo.go:98]: defer f.Close()

File: pkg/iacinventory/github_repo.go [hunk 16]
```diff
@@ -0,0 +1,184 @@
...
+	}
+	return nil
+}
+
+// GetMaster fetches and extracts the tarball of the master branch.
+func (g *GithubRepo) GetMaster(ctx context.Context, c *http.Client, username string, oauthToken string) error {
+	if username == "" || oauthToken == "" {
+		return fmt.Errorf("error fetching repository: credentials are not set")
+	}
+	req, _ := http.NewRequestWithContext(ctx, "GET", "https://api.github.com/repos/"+g.FullName+"/tarball/master", nil)
```
Comment (slshen) [inline pkg/iacinventory/github_repo.go:108]: master branch defaults to main for new repos now I believe.  If you use the google github api library it's easy to find the default branch.

File: pkg/iacinventory/github_repo.go [hunk 17]
```diff
@@ -0,0 +1,184 @@
...
+
+	// CloudformationDirs are directories that contain cloudformation files
+	CloudformationDirs map[string]bool // DEBT: lazy map implementation
+
+	// dir is where we've unarchived the tarball for analysis.
+	dir string
+}
+
+// GetTerraformDirs implements WalkFunc to search for directories that contain Terraform files.
+func (g *GithubRepo) GetTerraformDirs(path string, info os.FileInfo, err error) error {
```
Comment (slshen) [inline pkg/iacinventory/github_repo.go:41]: Don't export these funcs unless there's a need

File: cmd/iacinventorycmd/iacinventorycmd.go [hunk 18]
```diff
@@ -0,0 +1,53 @@
...
+	"github.com/soluble-ai/soluble-cli/pkg/iacinventory"
+	"github.com/soluble-ai/soluble-cli/pkg/options"
+	"github.com/spf13/cobra"
+)
+
+func Command() *cobra.Command {
+	var username string
+	var oauthToken string
+	var submit bool
+	opts := options.PrintClientOpts{}
```
Comment (slshen) [inline cmd/iacinventorycmd/iacinventorycmd.go:15]: So PrintClientOpts{} should be used when the command needs to print something back to the user.  It supports all sorts of things: tabular output, filtering, sorting, yaml/json output, etc.

ClientOpts{} just gives you an API client.

You can see the difference if you run the command with --help.  There's lots of printing-related options available for PrintClientOpts{}.

**Review by slshen — CHANGES_REQUESTED** [review]

File: pkg/iacinventory/iacinventory.go (hunk 6 above)
Comment (Eriner) [reply pkg/iacinventory/iacinventory.go:7]: Wasn't sure if I was going to need graceful handling of `^C` (deleting downloaded repos), but I haven't even implemented deleting downloaded repos...

File: pkg/iacinventory/github_repo.go (hunk 11 above)
Comment (Eriner) [reply pkg/iacinventory/github_repo.go:169]: It detects if the `io.CopyN` copied its maximum, 100MB. If it did, the repo was too big and needs to be skipped.

File: pkg/iacinventory/github_repo.go (hunk 14 above)
Comment (Eriner) [reply pkg/iacinventory/github_repo.go:21]: Kind of. GithubRepo is the data we can/do extract for GitHub. I explicitly separated the GitHub type from the generic `Repo` type. I manually construct the `Repo` from the `GithubRepo`. Rationale is that other providers (GitLab, Bitbucket, etc) down the road may not, for whatever reason, allow pulling out the requisite data. Or, the "name" of each may differ (`full_name` vs `name`), etc.

File: pkg/iacinventory/github_repo.go (hunk 11 above)
Comment (Eriner) [reply pkg/iacinventory/github_repo.go:169]: nevermind, i'm dumb. You are correct, fixed.

File: pkg/iacinventory/github_repo.go (hunk 9 above)
Comment (Eriner) [reply pkg/iacinventory/github_repo.go:162]: My statement above was incorrect. Changed to limit single file size to 1MB.

**Comment by Eriner** [comment]
It seems that `go-github` lacks functionality to fetch a repo's contents. I half-baked the transition to `go-github` before realizing this. I'm leaving the WIP functions commented out but pushing it anyway. I don't want to leave any `http.Get()s` in there if I don't have to, as it requires doing the header auth and whatnot manually anyway. Not being able to download the tarball with `go-github` just complicates the code and I'd prefer to just keep the `http.Request`s.

Will ctrl-f again after coffee.

File: pkg/iacinventory/github_repo.go (hunk 16 above)
Comment (Eriner) [reply pkg/iacinventory/github_repo.go:140]: Yep, leaving this one for once I figure out how to download the repo using `go-github`.

Comment (slshen) [reply pkg/iacinventory/github_repo.go:140]: https://pkg.go.dev/github.com/google/go-github/v32/github#RepositoriesService.GetArchiveLink

File: pkg/iacinventory/github_repo.go (hunk 14 above)
Comment (slshen) [reply pkg/iacinventory/github_repo.go:21]: Yeah, wouldn't worry about that.  If Repo is some denormalized structure with everything we could possibly send that's fine.

Comment (Eriner) [reply pkg/iacinventory/github_repo.go:24]: I created an Repo interface with the basic methods that are needed for any given repo for data extraction. This is probably what I should have done to begin with. Can add a `MarshalJSON` function that marshals the data to JSON as needed.

File: pkg/iacinventory/github.go [hunk 19]
```diff
@@ -0,0 +1,191 @@
...
+	}
+	if err := yaml.Unmarshal(yamlf, &ghConfig); err != nil {
+		retErr = fmt.Errorf("unable to parse gh config: %w", err)
+		return
+	}
+
+	n := jnode.FromMap(ghConfig)
+	username = n.Path("github.com").Path("user").String()
+	oauthToken = n.Path("github.com").Path("oauth_token").String()
+	log.Infof(username)
```
Comment (Eriner) [inline pkg/iacinventory/github.go:179]: This function is currently broken, and I have no idea why. The empty-string checks below never trip, despite the credentials not being properly read-in. No idea what is going on here.

Comment (slshen) [reply pkg/iacinventory/github.go:179]: String() returns JSON
GetString() returns a string value

File: pkg/iacinventory/github_repo.go (hunk 16 above)
Comment (Eriner) [reply pkg/iacinventory/github_repo.go]: Solved this by just omitting the ref from the download URL. This results in downloading the default branch.

File: pkg/iacinventory/github_repo.go (hunk 14 above)
Comment (Eriner) [reply pkg/iacinventory/github_repo.go:23]: Reverted that change. Just using the `GithubRepo` type for now. Will deal with the interface after we hook up another provider.

**Comment by Eriner** [comment]
Still lacks tests, but most comments have been addressed.

**Pull request merged**